### PR TITLE
Add synthetic canary construct

### PR DESCRIPTION
**Pull request by bkochendorfer**
# Goal

As part of the Pocket PagerDuty Workstream I am looking into adding general synthetic monitoring as part of our approach to reliability. This is an example of building a basic check and alarm into our shared `terraform-modules` repository. I think another approach could be to use NewRelic Synthetics which I will do in a separate PR so that we can compare approaches.

## Review focus
Looking for help on my `Typescript` in general. I am still new and it's been a very long time since I've used Javascript. Let me know if I made sane assumptions and approaches to accomplish the goal.

The way AWS requires a particular folder structure to get this to work is really bizarre.

## Outpus
I ran this in `dev` using the `example.ts` with some modifications and created this [Synthetic check](https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#synthetics:canary/detail/parser-check-syntheic) and this [Cloudwatch Alarm](https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#alarmsV2:alarm/parser-check-SyntheticsFailed-Alarm?)

## Todos
- [ ] Implement comparable logic in NewRelic Synthetics terraform provider

### CODE REVIEW

File: src/pocket/PocketSynthetic.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    port: string;
+    protocol: string;
+  },
+  check?: any;
+  environment: string;
+  artifactS3Location: string;
+  pagerDutyArn?: string;
+}
+
+const defaultMonitor = comile(`
```
Comment (kkyeboah) [inline src/pocket/PocketSynthetic.ts:26]:
```suggestion
const defaultMonitor = compile(`
```

File: src/pocket/PocketSynthetic.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
+import { Resource } from 'cdktf';
+import { Synthetics, IAM, CloudWatch } from '@cdktf/provider-aws';
+import { Construct } from 'constructs';
+import { ArchiveProvider, DataArchiveFile } from '@cdktf/provider-archive';
+
+// the exported type is called compile? I assume
```
Comment (kkyeboah) [inline src/pocket/PocketSynthetic.ts:6]: I just did a few google searches and couldn't find a built-in named `compile`. What happens when you name the variable `compile`?

File: src/pocket/PocketSynthetic.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+          this.config.config.protocol,
+          this.config.config.port,
+          this.config.config.path,
+        ) 
+    }
+
+    const indexPath = path.join(__dirname, 'synthetics', 'nodejs',
+        'node_modules', 'index.js');
+
+    fs.writeFile(indexPath, this.config.check, err => {
```
Comment (kkyeboah) [inline src/pocket/PocketSynthetic.ts:103]: suggestion: prefer `fs.writeFileSync` here to ensure the path is created before zipping the directory on like `110`

File: src/pocket/PocketSynthetic.spec.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+const config: PocketSyntheticProps = {
+  config: {
+    url: 'acme.getpocket.dev',
+    path: '/healthz',
+    port: '443',
+    protocol: 'https:'
+  },
+  environment: 'test',
+  artifactS3Location: 's3://not-a-real-thing'
```
Comment (kkyeboah) [inline src/pocket/PocketSynthetic.spec.ts:12]: should we create a default S3 bucket if a location is not provided?

**Review by kkyeboah — COMMENTED** [review]
I'm so hyped about this! thanks for adding a synthetic construct. I left a few comments.

Also, I think you will have linter issues, so safe to run `npm run lint:fix` on your commits.

File: src/pocket/PocketSynthetic.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        includeRequestBody: true,
+        includeResponseBody: false,
+        restrictedHeaders: ["authorization", "x-amz-security-token"],
+        continueOnHttpStepFailure: true
+    };
+
+    await synthetics.executeHttpStep('Verify synthetic check', {
+        ...commonProps,
+        path: '{3}',
+    }, validateSuccessfull, allConfig);
```
Comment (kkyeboah) [inline src/pocket/PocketSynthetic.ts:74]: nit: typo
```suggestion
    }, validateSuccessful, allConfig);
```

File: src/pocket/PocketSynthetic.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  pagerDutyArn?: string;
+}
+
+const defaultMonitor = comile(`
+var synthetics = require('Synthetics');
+
+const apiCanary = async function () {
+
+    // Handle validation for positive scenario
+    const validateSuccessfull = async function (res) {
```
Comment (kschelonka) [inline src/pocket/PocketSynthetic.ts:32]:
```suggestion
    const validateSuccessful = async function (res) {
```

File: src/pocket/PocketSynthetic.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    fs.writeFile(indexPath, this.config.check, err => {
+      if (err) {
+        console.log(err);
+        return
+      }
+    });
+
+    const zipFile = new DataArchiveFile(this, "synthetic-zip-file", {
+      type: "zip",
+      // TODO: hmm, maybe don't want to cause a diff evertyime
```
Comment (kschelonka) [inline src/pocket/PocketSynthetic.ts:112]: Will the terraform provider handle this automatically? e.g. checking the md5 against the md5 in state to see if it needs to be updated

File: src/pocket/synthetics/nodejs/node_modules/index.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+console.log('kelvin');
```
Comment (kschelonka) [inline src/pocket/synthetics/nodejs/node_modules/index.ts:1]: 😂

Comment (kkyeboah) [reply src/pocket/synthetics/nodejs/node_modules/index.ts:1]: I made it big

File: src/pocket/PocketSynthetic.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+`);
+
+export class PocketSyntheticCheck extends Resource {
+  constructor(
+    scope: Construct,
+    private name: string,
+    private config: PocketSyntheticProps
+  ) {
+    super(scope, name);
+    new ArchiveProvider(this, 'archive', {});
```
Comment (kkyeboah) [inline src/pocket/PocketSynthetic.ts:89]: remove provider init from constructs based on slack conversation: https://pocket.slack.com/archives/C03QVL4SU/p1637186040339800

File: src/pocket/PocketSynthetic.spec.ts (hunk 4 above)
Comment (bkochendorfer) [reply src/pocket/PocketSynthetic.spec.ts:12]: Having a bucket per check seems like a lot to me but I could be convinced otherwise!

**Comment by bkochendorfer** [comment]
Closing this in favor of the NewRelic path.

**Pull request closed without merging**